File: cvicenia/cv02/FormulaTest.java
import java.util.Map;
import java.util.HashMap;
import java.util.Set;
import java.util.HashSet;
import java.util.Arrays;

class ValCase {
    Map<String,Boolean> valuation;
    boolean result;

    ValCase(Map<String, Boolean> valuation, boolean result) {
        this.valuation = valuation;
        this.result    = result;
    }
}

class Case {
    Formula formula;
    String string;
    int deg;
    Set<String> vars;
    ValCase[] vals;

    Case(Formula f, String s, int d, Set<String> vars, ValCase[] vals){
        this.formula = f;
        this.string = s;
        this.deg = d;
        this.vars = vars;
        this.vals = vals;
    }
}

class Tester {
    int tested = 0;
    int passed = 0;

    public void compare(Object result, Object expected, String msg) {
        tested += 1;
        if (result.equals(expected)) {
            passed += 1;
        } else {
            System.out.println("    Failed: " + msg + ":");
            System.out.println("      got " + result + " expected " + expected);
        }
    }

    public void testToString(Case c) {
        compare(c.formula.toString(), c.string, "toString " + c.string);
    }

    public void testDeg(Case c) {
        compare(c.formula.deg(), c.deg, "deg " + c.string);
    }

    public void testVars(Case c) {
        compare(c.formula.vars(), c.vars, "vars " + c.string);
    }

    public void testIsSatisfied(Case c) {
        for (ValCase vc : c.vals) {
            System.out.println(" valuation " + vc.valuation + ":");
            compare(c.formula.isSatisfied(vc.valuation), vc.result, "isSatisfied");
        }
    }

    public void test(Formula formula, String string, ValCase[] cases) {
        System.out.println("Testing " + string);
        compare(formula.toString(), string, "toString");
        for (ValCase c : cases) {
            System.out.println(" valuation " + c.valuation + ":");
            compare(formula.isSatisfied(c.valuation), c.result, "isSatisfied");
        }
    }

    public void status() {
        System.out.println("TESTED " + tested + "\nPASSED " + passed);
        if (tested == passed) {
            System.out.println("OK");
        } else {
            System.out.println("ERROR");
        }
    }
}

/**
 *
 * @author shanki
 */
public class FormulaTest {

    /**
     * @param args the command line arguments
     */
    public static void main(String[] args) {
        Tester t = new Tester();

        Set<String> ss = new HashSet<String>(Arrays.asList("a", "b"));

        Map<String,Boolean> v1 = new HashMap<>();
        v1.put("a", true);

        Map<String,Boolean> v2 = new HashMap<>();
        v2.put("a", false);

        Case[] cases = new Case[] {
            new Case(new Variable("a"), "a", 0,
                new HashSet<String>(Arrays.asList("a")),
                new ValCase[] {
                new ValCase(v1, true), new ValCase(v2, false)
            }),

            new Case(new Negation(new Variable("a")), "-a", 1,
                new HashSet<String>(Arrays.asList("a")),
               new ValCase[] {
               new ValCase(v1, false), new ValCase(v2, true) 
            }),
        };

        System.out.println("Testing toString");
        for (Case c : cases) {
            t.testToString(c);
        }

        System.out.println("Testing deg");
        for (Case c : cases) {
            t.testDeg(c);
        }

        System.out.println("Testing vars");
        for (Case c : cases) {
            t.testVars(c);
        }

        System.out.println("Testing isSatisfied");
        for (Case c : cases) {
            t.testIsSatisfied(c);
        }

        Map<String,Boolean> a1 = new HashMap<>();
        a1.put("a", false);
        a1.put("b", false);

        Map<String, Boolean> a2 = new HashMap<>();
        a2.put("a", false);
        a2.put("b", true);

        Map<String, Boolean> a3 = new HashMap<>();
        a3.put("a", true);
        a3.put("b", false);

        Map<String, Boolean> a4 = new HashMap<>();
        a4.put("a", true);
        a4.put("b", true);

        t.test(new Conjunction( new Formula[] { new Variable("a"), new Variable("b") } ), "(a&b)", new ValCase[] {
        new ValCase(a1, false), new ValCase(a2, false), new ValCase(a3, false), new ValCase(a4, true)
        });

        t.test(new Disjunction( new Formula[] { new Variable("a"), new Variable("b") } ), "(a|b)", new ValCase[] {
        new ValCase(a1, false), new ValCase(a2, true), new ValCase(a3, true), new ValCase(a4, true)
        });

        t.test(new Implication( new Variable("a"), new Variable("b") ), "(a->b)", new ValCase[] {
        new ValCase(a1, true), new ValCase(a2, true), new ValCase(a3, false), new ValCase(a4, true)
        });

        t.test(new Equivalence( new Variable("a"), new Variable("b") ), "(a<->b)", new ValCase[] {
        new ValCase(a1, true), new ValCase(a2, false), new ValCase(a3, false), new ValCase(a4, true)
        });

        t.test(new Disjunction(new Formula[] {
            new Negation(new Implication(new Variable("a"), new Variable("b"))),
            new Negation(new Implication(new Variable("b"), new Variable("a")))
        }), "(-(a->b)|-(b->a))", new ValCase[] {
          new ValCase(a1, false), new ValCase(a2, true), new ValCase(a3, true), new ValCase(a4, false)  
        });

        Map<String,Boolean> b1 = new HashMap<>();
        b1.put("a", false);
        b1.put("b", false);
        b1.put("c", false);

        Map<String,Boolean> b2 = new HashMap<>();
        b2.put("a", true);
        b2.put("b", false);
        b2.put("c", false);

        Map<String,Boolean> b3 = new HashMap<>();
        b3.put("a", false);
        b3.put("b", true);
        b3.put("c", false);

        Map<String,Boolean> b4 = new HashMap<>();
        b4.put("a", true);
        b4.put("b", true);
        b4.put("c", false);

        Map<String,Boolean> b5 = new HashMap<>();
        b5.put("a", false);
        b5.put("b", false);
        b5.put("c", true);

        Map<String,Boolean> b6 = new HashMap<>();
        b6.put("a", true);
        b6.put("b", false);
        b6.put("c", true);

        Map<String,Boolean> b7 = new HashMap<>();
        b7.put("a", false);
        b7.put("b", true);
        b7.put("c", true);

        Map<String,Boolean> b8 = new HashMap<>();
        b8.put("a", true);
        b8.put("b", true);
        b8.put("c", true);

        t.test(new Conjunction(new Formula[] {
            new Implication(new Variable("a"), new Variable("b")),
            new Implication(new Negation(new Variable("a")), new Variable("c"))
        }), "((a->b)&(-a->c))", new ValCase[] {
            new ValCase(b1, false), new ValCase(b2, false), new ValCase(b3, false), 
            new ValCase(b4, true), new ValCase(b5, true), new ValCase(b6, false),
            new ValCase(b7, true), new ValCase(b8, true)
        });

        t.test(new Equivalence(
                new Conjunction(new Formula[] { new Variable("a"), new Negation(new Variable("b")) }),
                new Disjunction(new Formula[] { new Variable("a"), new Implication( new Variable("b"), new Variable("a")) })
                ), "((a&-b)<->(a|(b->a)))", new ValCase[] {
                    new ValCase(a1, false), new ValCase(a2, true), new ValCase(a3, true), new ValCase(a4, false)
                }
        );

        {
            System.out.println("Testing Negation.originalFormula");
            Formula a = new Variable("a");
            Formula na = new Negation(a);
            Negation nna = new Negation(na);
            t.compare(nna.originalFormula(), na, "Negation.originalFormula");
        }

        {
            System.out.println("Testing Implication rightSide / leftSide");
            Formula a = new Variable("a");
            Formula b = new Variable("b");
            Formula na = new Negation(a);
            Implication nab = new Implication(na, b);
            t.compare(nab.leftSide(), na, "Implication.leftSide");
            t.compare(nab.rightSide(), b, "Implication.rightSide");
        }

        {
            System.out.println("Testing Equivalence rightSide / leftSide");
            Formula a = new Variable("a");
            Formula b = new Variable("b");
            Formula na = new Negation(a);
            Equivalence nab = new Equivalence(na, b);
            t.compare(nab.leftSide(), na, "Equivalence.leftSide");
            t.compare(nab.rightSide(), b, "Equivalence.rightSide");
        }
        t.status();
    }
}

File: cvicenia/cv02/formulaTest.py
#!/bin/env python3

"""
  Testovaci program pre kniznicu formula.py
"""

from formula import Variable, Negation, Conjunction, Disjunction, Implication, Equivalence

class Tester(object):
    def __init__(self):
        self.tested = 0
        self.passed = 0
    def compare(self, result, expected, msg):
        self.tested += 1
        if result == expected:
            self.passed += 1
        else:
            print("    Failed: %s:" %  msg)
            print("      got %s  expected %s" % (repr(result), repr(expected)))

    def testToString(self, formula, string):
        self.compare(formula.toString(), string, "toString %s" % string)

    def testDeg(self, formula, string, deg):
        self.compare(formula.deg(), deg, "deg %s" % string)

    def testVars(self, formula, string, vs):
        fvars = formula.vars()
        self.compare(isinstance(fvars, (set, frozenset)), True, "vars is a set %s" % string)
        self.compare(frozenset(fvars), frozenset(vs), "vars %s" % string)

    def testIsSatisfied(self, formula, string, cases):
        for valuation, result in cases:
            v = " valuation %s" % (repr(valuation),)
            self.compare(formula.isSatisfied(valuation), result, "isSatisfied %s %s" % (string, v))

    def status(self):
        print("TESTED %d" % (self.tested,))
        print("PASSED %d" % (self.passed,))
        if self.tested == self.passed:
            print("OK")
        else:
            print("ERROR")

t = Tester()

valuations2 = [{ 'a': False, 'b': False },
               { 'a': False, 'b': True  },
               { 'a': True , 'b': False },
               { 'a': True , 'b': True  }]

valuations3 = [{ 'a': False, 'b': False, 'c': False },
               { 'a': True , 'b': False, 'c': False },
               { 'a': False, 'b': True , 'c': False },
               { 'a': True , 'b': True , 'c': False },
               { 'a': False, 'b': False, 'c': True  },
               { 'a': True , 'b': False, 'c': True  },
               { 'a': False, 'b': True , 'c': True  },
               { 'a': True , 'b': True , 'c': True  }]

testFormulas = [
    (
        Variable('a'), 'a', 0, ['a'],
        [
            ({'a':True}, True),
            ({'a':False}, False),
        ]),

    (
        Negation(Variable('a')), '-a', 1, ['a'],
        [
            ({'a':True}, False),
            ({'a':False}, True),
        ]),



    (
        Conjunction( [ Variable('a'), Variable('b') ] ),
        '(a&b)',
        1,
        ['a','b'],
        [
            (valuations2[0], False),
            (valuations2[1], False),
            (valuations2[2], False),
            (valuations2[3], True),
        ]),

    (
        Disjunction( [ Variable('a'), Variable('b') ] ),
        '(a|b)',
        1,
        ['a','b'],
        [
            (valuations2[0], False),
            (valuations2[1], True),
            (valuations2[2], True),
            (valuations2[3], True),
        ]),

    (
        Implication( Variable('a'), Variable('b') ),
        '(a->b)',
        1,
        ['a','b'],
        [
            (valuations2[0], True),
            (valuations2[1], True),
            (valuations2[2], False),
            (valuations2[3], True),
        ]),

    (
        Equivalence( Variable('a'), Variable('b') ),
        '(a<->b)',
        1,
        ['a','b'],
        [
            (valuations2[0], True),
            (valuations2[1], False),
            (valuations2[2], False),
            (valuations2[3], True),
        ]),

    (
        Disjunction([
            Negation(Implication(Variable('a'),Variable('b'))),
            Negation(Implication(Variable('b'),Variable('a')))
        ]),
        '(-(a->b)|-(b->a))',
        5,
        ['a','b'],
        [
            (valuations2[0], False),
            (valuations2[1], True),
            (valuations2[2], True),
            (valuations2[3], False),
        ]),


    (
        Conjunction([
            Implication(Variable('a'),Variable('b')),
            Implication(Negation(Variable('a')),Variable('c'))
        ]),
        '((a->b)&(-a->c))',
        4,
        ['a','b','c'],
        [
            (valuations3[0], False),
            (valuations3[1], False),
            (valuations3[2], False),
            (valuations3[3], True),
            (valuations3[4], True),
            (valuations3[5], False),
            (valuations3[6], True),
            (valuations3[7], True),
        ]),

    (
        Equivalence(
            Conjunction([
                Variable('a'),
                Negation(Variable('b'))
            ]),
            Disjunction([
                Variable('a'),
                Implication(
                    Variable('b'),
                    Variable('a')
                )
            ])
        ),
        '((a&-b)<->(a|(b->a)))',
        5,
        ['a','b'],
        [
            (valuations2[0], False),
            (valuations2[1], True),
            (valuations2[2], True),
            (valuations2[3], False),
        ]),

    (
        Conjunction([Variable('a'), Variable('a'), Variable('a'), Variable('a')]),
        '(a&a&a&a)',
        1,
        ['a'],
        [
            ({'a':True}, True),
            ({'a':False}, False),
        ]),

    (
        Disjunction([Variable('a'), Variable('a'), Variable('a'), Variable('a')]),
        '(a|a|a|a)',
        1,
        ['a'],
        [
            ({'a':True}, True),
            ({'a':False}, False),
        ]),

    (
        Conjunction([Variable('a'), Variable('b'), Variable('c')]),
        '(a&b&c)',
        1,
        ['a','b','c'],
        [
            (valuations3[0], False),
            (valuations3[1], False),
            (valuations3[2], False),
            (valuations3[3], False),
            (valuations3[4], False),
            (valuations3[5], False),
            (valuations3[6], False),
            (valuations3[7], True),
        ]),

    (
        Disjunction([Variable('a'), Variable('b'), Variable('c')]),
        '(a|b|c)',
        1,
        ['a','b','c'],
        [
            (valuations3[0], False),
            (valuations3[1], True),
            (valuations3[2], True),
            (valuations3[3], True),
            (valuations3[4], True),
            (valuations3[5], True),
            (valuations3[6], True),
            (valuations3[7], True),
        ]),
]

print("Testing toString")
for f,s,d,vars,vals in testFormulas:
    t.testToString(f, s)

print("Testing isSatisfied")
for f,s,d,vars,vals in testFormulas:
    t.testIsSatisfied(f, s, vals)

print("Testing deg")
for f,s,d,vars,vals in testFormulas:
    t.testDeg(f, s, d)

print("Testing vars")
for f,s,d,vars,vals in testFormulas:
    t.testVars(f, s, vars)

print("Testing subf")
ca = Variable('a')
cnb = Variable('b')
cn = Negation(cnb)
c = Conjunction([ca, cn])

da = Variable('a')
dib = Variable('b')
dia = Variable('a')
di = Implication(dib, dia)
d = Disjunction([da, di])
f = Equivalence(c, d)

t.compare(f.subf(), [c, d], "subf (Eq)")

t.compare(f.subf()[0].subf(), [ca, cn], "subf [0] (Conj)")
t.compare(f.subf()[0].subf()[0].subf(), [], "subf [0][0] (Var)")
t.compare(f.subf()[0].subf()[1].subf(), [cnb], "subf [0][1] (Neg)")
t.compare(f.subf()[0].subf()[1].subf()[0].subf(), [], "subf [0][1][0] (Var)")

t.compare(f.subf()[1].subf(), [da, di], "subf [1] (Disj)")
t.compare(f.subf()[1].subf()[0].subf(), [], "subf [1][0] (Var)")
t.compare(f.subf()[1].subf()[1].subf(), [dib, dia], "subf [1][1] (Impl)")
t.compare(f.subf()[1].subf()[1].subf()[0].subf(), [], "subf [1][1][0] (Var)")
t.compare(f.subf()[1].subf()[1].subf()[1].subf(), [], "subf [1][1][0] (Var)")

a = Variable('a')
b = Variable('b')
c = Variable('c')
d = Variable('a')
conj = Conjunction([a, b, c, d])
t.compare(conj.subf(), [a, b, c, d], "subf Conj3")

print("Testing Variable.name")
a = Variable('a')
b = Variable('b')
t.compare(a.name(), 'a', "Variable.name")
t.compare(b.name(), 'b', "Variable.name")


print("Testing Negation.originalFormula")
a = Variable('a')
na = Negation(a)
nna = Negation(na)
t.compare(nna.originalFormula() is na, True, "Negation.originalFormula")

print("Testing Implication rightSide / leftSide")
a = Variable('a')
b = Variable('b')
na = Negation(a)
nab = Implication(na, b)
t.compare(nab.leftSide() is na, True, "Implication.leftSide")
t.compare(nab.rightSide() is b, True, "Implication.rightSide")

print("Testing Equivalence rightSide / leftSide")
a = Variable('a')
b = Variable('b')
na = Negation(a)
nab = Equivalence(na, b)
t.compare(nab.leftSide() is na, True, "Equivalence.leftSide")
t.compare(nab.rightSide() is b, True, "Equivalence.rightSide")

print("Testing equals")
t.compare(Variable('a').equals(Variable('a')), True, 'equals')
t.compare(Variable('a').equals(Variable('b')), False, 'equals')

a = Equivalence(
        Conjunction([
            Variable('a'),
            Negation(Variable('b'))
        ]),
        Disjunction([
            Variable('a'),
            Implication(
                Variable('b'),
                Variable('a')
            )
        ])
    )
b = Equivalence(
        Conjunction([
            Variable('a'),
            Negation(Variable('b'))
        ]),
        Disjunction([
            Variable('a'),
            Implication(
                Variable('b'),
                Variable('a')
            )
        ])
    )
c = Equivalence(
        Disjunction([
            Variable('a'),
            Negation(Variable('b'))
        ]),
        Disjunction([
            Variable('a'),
            Implication(
                Variable('b'),
                Variable('a')
            )
        ])
    )
t.compare(a.equals(a), True, 'equals')
t.compare(a.equals(b), True, 'equals')
t.compare(a.equals(c), False, 'equals')

print("Testing substitute")
w = Conjunction([Variable('a'), Negation(Variable('b'))])
r = Disjunction([Variable('a'), Negation(Variable('b'))])
s = a.substitute(w, r)
t.compare(s.equals(c), True, 'substitute %s for %s in %s -> %s' % (w.toString(), r.toString(), a.toString(), s.toString()))
t.compare(s is not a, True, 'substite creates new copy')

t.compare(s.leftSide().equals(r), True, 'substitute replaces with equal formula')
t.compare(s.leftSide() is not r, True, 'substitute replaces with new copy')

r2 = r.substitute(Variable('b'), Variable('c'))
t.compare(r2.equals(Disjunction([Variable('a'), Negation(Variable('c'))])), True,
        'substitute b for c in %s -> %s' % (r.toString(), r2.toString()))
t.compare(r2.equals(r), False, 'substitute creates new')


t.status()

# vim: set sw=4 ts=4 sts=4 sw :
